### PR TITLE
Add `JobBlock` implementation recipe

### PR DESCRIPTION
**Pull request by zzstoatzz**
## Description

What's cooking? 
If this PR references an issue, please attach to the existing issue.

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## New Recipe Checklist

- [ ] My PR is in the format of `Add <project-name> recipe`
- [ ] My recipe is reproducible and explains everything needed to run successfully. If my code has external dependencies, I make mention of them.
- [ ] My code is easily understandable and/or well-commented.
- [ ] If my recipe requires a new category (e.g. creating a monitoring/ folder in devops/), I have encapsulated my project within its own subfolder so that others can add their own recipes as well.
- [ ] If my recipe uses Prefect < 2.0, my code is within the prefect-v1-legacy/ folder.

### CODE REVIEW

**Comment by ahuang11** [comment]
I'm personally not a big fan of Foo/Bar examples because they don't really mean anything and they're not relatable (to me anyways). Here's a blog post I randomly found: https://www.techyourchance.com/never-use-foo-bar-baz/

I prefer if this recipe was based off a real life scenario.

File: extending-prefect/blocks/job_block.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    """
+
+    job_run = await task(foo_job.trigger.aio)(foo_job)
+
+    await task(job_run.wait_for_completion.aio)(job_run)
+
+    return await task(job_run.fetch_result.aio)(job_run)
+
+
+# and now we can run our flows - since Foo is definitely a real service!
```
Comment (ahuang11) [inline extending-prefect/blocks/job_block.py:176]: Maybe this can be in a separate file

File: extending-prefect/blocks/job_block.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+# now let's write a `FooJobRun` class that represents a specific run of a `FooJob`
+class FooJobRun(JobRun):
+    """A specific run of a `FooJob` block."""
+
+    def __init__(self, foo_job, job_id):
+        self.foo_job = foo_job
+        self.job_id = job_id
+        self._terminal_states = {"complete", "failed", "cancelled"}
+
+    @sync_compatible
```
Comment (ahuang11) [inline extending-prefect/blocks/job_block.py:75]: Maybe add a comment about what's sync_compatible

**Review by ahuang11 — COMMENTED** [review]
I like the idea of having extending-prefect/blocks; thanks for doing this!

**Comment by zzstoatzz** [comment]
Fair point about `Foo` being non-realistic, I guess I just didn't want to step on the toes of any real collections / blocks that we're writing and confuse anyone 

just wanted to put this out here so we could start talking about what should be here

**Review by desertaxle — COMMENTED** [review]
I think it's a great idea to document how to create different types of blocks, but I lean towards adding this to the Prefect docs under the Blocks concept page. We can sketch out a structure for how to document how to use and how to implement the various block interfaces.

**Pull request closed without merging**